plugin/token/approle/approle_test.go

```diff
@@ -0,0 +1,229 @@
+		}
+		data, _ := ioutil.ReadFile("testdata/new_token.json")
+		w.Write(data)
+	}
+	config, listener := testHTTPServer(t, http.HandlerFunc(handler))
+	defer listener.Close()
+
+	tlsConfig := api.TLSConfig{Insecure: true}
+	config.ConfigureTLS(&tlsConfig)
+
+	client, vaultErr := api.NewClient(config)
+	if vaultErr != nil {
+		t.Errorf("Can't connect to vault test server\n#{err}")
+	}
+
+	r := NewRenewer(client, roleId, secretId, ttl)
+	err := r.NewToken(noContext)
+
+	if err != nil {
+		t.Errorf("ERROR: %v", err)
+	}
+	if r.client.Token() != newToken {
+		t.Errorf("ERROR: expected token %v, got token %v", newToken, r.client.Token())
+	}
+
+	t.Parallel()
+}
+
+// Test logic when returned TTL is higher than requested
+func TestVaultApproleRenewHigherTTL(t *testing.T) {
+	handler := func(w http.ResponseWriter, req *http.Request) {
+		var data []byte
+		if req.URL.Path == "/v1/auth/token/renew" {
+			data, _ = ioutil.ReadFile("testdata/renew_higher_ttl.json")
+		} else if req.URL.Path == "/v1/auth/approle/login" {
+			data, _ = ioutil.ReadFile("testdata/new_token.json")
+		} else{
+			t.Errorf("Invalid path, %v", req.URL.Path)
+		}
+
+		w.Write(data)
+	}
+	config, listener := testHTTPServer(t, http.HandlerFunc(handler))
+	defer listener.Close()
+
+	tlsConfig := api.TLSConfig{Insecure: true}
+	config.ConfigureTLS(&tlsConfig)
+
+	client, vaultErr := api.NewClient(config)
+	client.SetToken(renewToken)
+	if vaultErr != nil {
+		t.Errorf("Can't connect to vault test server\n#{err}")
+	}
+
+	r := NewRenewer(client, roleId, secretId, ttl)
+	err := r.Renew(noContext)
+
+	if err != nil {
+		t.Errorf("ERROR: %v", err)
+	}
+	if r.client.Token() != renewToken {
+		t.Errorf("ERROR: expected token %v, got token %v", renewToken, r.client.Token())
+	}
+
+	t.Parallel()
+}
+
+// Test logic when returned TTL is lower than requested
+// Expect a new token to be generated
+func TestVaultApproleRenewLowerTTL(t *testing.T) {
+	handler := func(w http.ResponseWriter, req *http.Request) {
+		var data []byte
+		if req.URL.Path == "/v1/auth/token/renew" {
+			data, _ = ioutil.ReadFile("testdata/renew_lower_ttl.json")
+		} else if req.URL.Path == "/v1/auth/approle/login" {
+			data, _ = ioutil.ReadFile("testdata/new_token.json")
+		} else{
+			t.Errorf("Invalid path, %v", req.URL.Path)
+		}
+
+		w.Write(data)
+	}
+	config, listener := testHTTPServer(t, http.HandlerFunc(handler))
+	defer listener.Close()
+
+	tlsConfig := api.TLSConfig{Insecure: true}
+	config.ConfigureTLS(&tlsConfig)
+
+	client, vaultErr := api.NewClient(config)
+	client.SetToken(renewToken)
+	if vaultErr != nil {
+		t.Errorf("Can't connect to vault test server\n#{err}")
+	}
+
+	r := NewRenewer(client, roleId, secretId, ttl)
+	err := r.Renew(noContext)
+
+	if err != nil {
+		t.Errorf("ERROR: %v", err)
+	}
+	if r.client.Token() != newToken {
+		t.Errorf("ERROR: expected token %v, got token %v", newToken, r.client.Token())
+	}
+
+	t.Parallel()
+}
+
+func testHTTPServer(
```

Comment:
Go has an `httptest` package with a test server. This is preferable because it chooses a random port at runtime and will not conflict with existing ports already in use.

I recommend looking at how we use the test server, and trying to more closely mirror the coding style (no need for a helper function, for example).
https://github.com/drone/drone-vault/blob/master/plugin/plugin_test.go#L31:L34

Reply:
OK can do.  I lifted that logic directly from the Vault repo test logic and tried to mirror testing like they did.  I can see where you would want things consistent in a repo.  I'll make those updates